### PR TITLE
Improve error display for messages sent from insecure devices

### PR DESCRIPTION
**Pull request by uhoreg**
https://github.com/element-hq/matrix-react-sdk/pull/92 provides a mode in which we will reject:
 - messages sent from devices which their owners have not cross-signed
 - messages sent from users who we had previously verified, but who have now rotated their identity.

This PR improves the display of both kinds of message.

The former is shown like this:

image

The latter is shown like this:

image

This is based on the designs at https://github.com/element-hq/element-meta/issues/2523, but we don't yet show the content of the event (that being the point of https://github.com/element-hq/element-meta/issues/2523).

## Checklist

-   [x] Tests written for new code (and old code if feasible).
-   [ ] New or updated `public`/`exported` symbols have accurate [TSDoc](https://tsdoc.org/) documentation.
-   [x] Linter and other CI checks pass.
-   [ ] I have licensed the changes to Element by completing the [Contributor License Agreement (CLA)](https://cla-assistant.io/element-hq/matrix-react-sdk)

### CODE REVIEW

**Comment by uhoreg** [comment]
Ah.  Forgot that SonarCloud doesn't take into account the end-to-end tests, so I'll have to add jest tests.

File: src/components/views/messages/DecryptionFailureBody.tsx [hunk 1]
```diff
@@ -33,15 +33,33 @@ function getErrorMessage(mxEvent: MatrixEvent, isVerified: boolean | undefined):
 
         case DecryptionFailureCode.HISTORICAL_MESSAGE_USER_NOT_JOINED:
             return _t("timeline|decryption_failure|historical_event_user_not_joined");
+
+        case DecryptionFailureCode.SENDER_IDENTITY_PREVIOUSLY_VERIFIED:
+            return _t("timeline|decryption_failure|sender_identity_previously_verified");
+
+        case DecryptionFailureCode.UNSIGNED_SENDER_DEVICE:
+            // TODO: event should be hidden instead of showing this error (only
```
Comment (uhoreg) [inline src/components/views/messages/DecryptionFailureBody.tsx:41]: I think this needs to be done by checking the decryption failure code in `EventTile.tsx`, and if it is this code, then render an empty tile?

Comment (uhoreg) [inline src/components/views/messages/DecryptionFailureBody.tsx:41]: I think this needs to be done by checking the decryption failure code in `EventTile.tsx`, and if it is this code, then render an empty tile?

Comment (richvdh) [reply src/components/views/messages/DecryptionFailureBody.tsx:41]: Probably, but I'd like us to kick this down the road for a bit. Let's display the error for now and think about hiding them later.

File: res/css/views/messages/_DecryptionFailureBody.pcss [hunk 2]
```diff
@@ -10,3 +10,7 @@ Please see LICENSE files in the repository root for full details.
     color: $secondary-content;
     font-style: italic;
 }
+
+.mx_DecryptionFailureVerifiedIdentityChanged {
+    color: red;
```
Comment (richvdh) [inline res/css/views/messages/_DecryptionFailureBody.pcss:15]: I feel like we shouldn't be hardcoding colors like this

File: src/components/views/messages/DecryptionFailureBody.tsx
Comment (richvdh) [inline src/components/views/messages/DecryptionFailureBody.tsx:1]: could we maybe move the changes to DecryptionFailureBody into a separate PR?

**Review by richvdh — CHANGES_REQUESTED** [review]

File: src/settings/Settings.tsx [hunk 3]
```diff
@@ -316,6 +317,16 @@ export const SETTINGS: { [setting: string]: ISetting } = {
...
+    "feature_invisible_crypto": {
+        isFeature: true,
+        labsGroup: LabGroup.Encryption,
+        controller: new InvisibleCryptoController(),
+        displayName: _td("labs|invisible_crypto"),
+        description: _td("labs|invisible_crypto_description"),
+        supportedLevels: LEVELS_DEVICE_ONLY_SETTINGS_WITH_CONFIG_PRIORITISED,
+        supportedLevelsAreOrdered: true,
+        default: false,
+    },
```
[the remark this answers is not in the document]
Comment (uhoreg) [reply src/settings/Settings.tsx:329]: Yeah, I wasn't sure how best to represent this.  Having two separate toggles for three levels seems wrong (e.g. what do we do if both "invisible" and "transition" are enabled).  I think that ideally, this would be a dropdown or something that would allow us to select one of the three levels, but I don't how to do that in settings, or if it's even possible with feature flags.  I think that if it isn't possible for feature flags, my second preference would be for the feature flag to be to enable "new cryptography modes" or some such, which would enable a selector somewhere else to select the mode.  But, basically, I didn't have time to sufficiently investigate.

File: src/components/views/messages/DecryptionFailureBody.tsx
Comment (uhoreg) [reply src/components/views/messages/DecryptionFailureBody.tsx:1]: So, one PR for the config setting, and one PR for the display changes?  Possibly, but I suspect that the config setting PR will have a lack of code worth testing, which would cause SonarCloud to complain.  I'll rewrite this PR so that it has two commits - one for the config setting and one for the display changes, so that it can be split into two PRs.

Comment (uhoreg) [reply src/components/views/messages/DecryptionFailureBody.tsx:1]: I discovered some tests for settings, and added one for invisible crypto.  So maybe SonarCloud would be happy with a PR for just the setting.  But I've run out of time to split it into a separate PR.  But it is now in two commits.

**Comment by uhoreg** [comment]
A couple of  issue with this PR:
-  we don't retry decryption after we've verified a user.  I've created https://github.com/element-hq/matrix-react-sdk/issues/58 to track the issue.
- we don't have any special handling for the display in the roomlist message preview, or the thread preview in the main timeline.  So we end up with things like 
image
I've asked for designs for these at https://github.com/element-hq/element-meta/issues/2523#issuecomment-2358716334 so we may want to just wait until we implement https://github.com/element-hq/element-meta/issues/2523 to address this.

File: src/settings/Settings.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (dbkr) [reply src/settings/Settings.tsx:329]: Yeah, I suspect probably a custom labs flag that's actually a drop down 9and maybe not actually a labs flag but manually put into the labs section?) That or a labs flag to enable another setting in the crypto section, although that risks people not realising they've turned the labs flag on.

Either way, would you want this PR is as it is with no option for transition mode or do we want to wait?

**Review by dbkr — COMMENTED** [review]
Seems plausible otherwise

**Comment by uhoreg** [comment]
We should check whether this get counted in the PostHog stats.  I suspect it does, and we should probably track it as a separate code.

File: src/settings/Settings.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (richvdh) [reply src/settings/Settings.tsx:329]: see https://github.com/element-hq/element-meta/issues/2536#issuecomment-2368346133 for questions about whether we actually need a tristate

[the remark this answers is not in the document]
Comment (richvdh) [reply src/settings/Settings.tsx:329]: After discussion: we do not need two settings here. But also, this whole PR needs to block on https://github.com/matrix-org/matrix-js-sdk/issues/4428.

**Comment by richvdh** [comment]
> We should check whether this get counted in the PostHog stats. I suspect it does, and we should probably track it as a separate code.

Filed as https://github.com/element-hq/element-meta/issues/2541

**Comment by richvdh** [comment]
Closing in favour of #93
